toffy/reorg.py

```diff
@@ -0,0 +1,145 @@
+    os.makedirs(output_dir)
+
+    # loop over each run
+    for run in run_folders:
+        run_path = os.path.join(cohort_dir, run)
+
+        fovs = io_utils.list_folders(run_path)
+        for fov in fovs:
+            shutil.move(os.path.join(run_path, fov),
+                        os.path.join(output_dir, run + '_' + fov))
+
+        shutil.rmtree(run_path)
+
+
+def rename_fov_dirs(json_run_path, default_run_dir, output_run_dir=None):
+    """Renames FOV directories with default_name to have custom_name sourced from the run JSON file
+
+    Args:
+        json_run_path (str): path to the JSON run file which contains the custom name values
+        default_run_dir (str): directory containing default named FOVs
+        output_run_dir (str): directory for renamed FOVs. If None, changed in place
+
+    Raises:
+        KeyError: issue reading keys from the JSON file
+        ValueError: there exist default named directories that are not described in the run file
+        UserWarning: not all custom names from the run file have an existing directory
+
+        """
+
+    io_utils.validate_paths(json_run_path)
+    io_utils.validate_paths(default_run_dir)
+
+    # check that new_dir doesn't already exist
+    if output_run_dir is not None:
+        if os.path.exists(output_run_dir):
+            raise ValueError(f"The new directory supplied already exists: {output_run_dir}")
+
+    with open(json_run_path) as file:
+        run_metadata = json.load(file)
+
+    # check for missing or duplicate fov names
+    run_metadata = rename_missing_fovs(run_metadata)
+    run_metadata = rename_duplicate_fovs(run_metadata)
+
+    # retrieve custom names and number of scans for each fov, construct matching default names
+    fov_scan = dict()
+    for fov in run_metadata.get('fovs', ()):
+        custom_name = fov.get('name')
+        run_order = fov.get('runOrder', -1)
+        scans = fov.get('scanCount', -1)
+        if run_order * scans < 0:
+            raise KeyError(f"Could not locate keys in {run_path}")
+
+        # fovs with multiple scans have scan number specified
+        if scans > 1:
+            for scan in range(1, scans + 1):
+                default_name = f'fov-{run_order}-scan-{scan}'
+                fov_scan[default_name] = f'{custom_name}-{scan}'
+        else:
+            default_name = f'fov-{run_order}-scan-{scans}'
+            fov_scan[default_name] = custom_name
+
+    # retrieve current default directory names, check if already renamed
+    old_dirs = io_utils.list_folders(default_run_dir, 'fov')
+    if len(old_dirs) == 0:
+        raise ValueError(f"All FOV folders in {default_run_dir} have already been renamed")
+
+    # check if custom fov names & scan counts match the number of existing default directories
+    verify_in_list(warn=True, fovs_in_run_file=list(fov_scan.keys()),
+                   existing_fov_folders=old_dirs)
+    verify_in_list(existing_fov_folders=old_dirs, fovs_in_run_file=list(fov_scan.keys()))
```

Comment:
This works but also uses `verify_in_list` awkwardly (first to warn the user of invalid values, then to trigger the error). Is there anyway we can simply leverage the error statement without having to make two `verify_in_list` calls.

Reply:
It's because we want to warn about FOVs in the JSON that aren't found in directory, but throw an error for FOVs in the directory that are not found in the JSON